### PR TITLE
event_spamblock: Fixed forced comment moderation 0 -> moderate all

### PR DESCRIPTION
**Pull request by stephanbrunker**
for the option: "Moderate comments after x days" the function was missing that for a value of 0, all comments are going to be moderated. I fixed this.

### CODE REVIEW

**Comment by onli** [comment]
The code is explicitly doing the contrary: 0 means no moderation. That also matches the (german) description:

> 0 bedeutet, dass keine automatische Moderation erzeugt wird.

We often need a way to have a null value that disables an option.

When wanting to moderate all comments this can be set in the personal settings as a default setting for the entry editor.

Is this maybe more about wanting a way to force-moderate all comments, of new and existing entries?

**Comment by stephanbrunker** [comment]
Oh … if 0 means no moderation, then that setting is not consistent to the captcha setting slightly above where 0 means "force captchas". And yes, the goal is to force-moderate all existing entries. If the value is 0.0000116, this means after 1 second, so logically 0 means immediately. To turn the auto-moderation off, a value of "infinite" would be correct, so perhaps a -1 ? Or an extra option to turn it off, like it is for the captchas?

**Comment by onli** [comment]
Right, that is not consistent at all.

> Or an extra option to turn it off, like it is for the captchas?

That sounds best to me. Then it would be like the captcha setting. *Auto-moderate* comments yes/no, then the time until auto-moderation, default being 0.

Maybe you can add this to this PR? The default should be off.

But then we should also have an upgrader task, that sets the new boolean *Auto-moderate* option to false if the interval is 0 now, and to true if it is a higher number.

**Comment by stephanbrunker** [comment]
I can do parts of this task, like the changes in the event_spamblock.php and the DE and EN language files. What I cannot do are the other languages and the upgrader/database operations. 

This means I add 'forcemoderation_auto' into introspect->propbag ('configuration') and a new case at introspect_config_item? Is the field in the config database then created automatically with this 'propbag' or is that something which has to be done manually first?

**Comment by stephanbrunker** [comment]
I did what I can do - the event_spamblock.php and EN and DE language files.

**Comment by stephanbrunker** [comment]
This is much like poking in the mist, but I think it could be mostly correct. I am not sure about the version number, but the function should activate the moderation if a non-zero value is set. If it is set to zero and thus the moderation is off, the default value 'false' doesn't have to be changed.

**Comment by onli** [comment]
The logic looks fine, thank you. Did you test the function itself? I'm surprised that `serendipity_db_query` would return a straight value and not an array.

**Comment by stephanbrunker** [comment]
No, as I wrote I honestly don't have a clue of what I coded (outside of SQL and PHP in general). I just made it up by altering one of the existing functions and looked up the names of the db fields. I have an active installation of serendipity, but not a test environment, so I couldn't test it. (I also don't know which version number I have to insert in the upgrader.inc.php) If serendipity_db_query returns an array, then I have to copy&paste that snippet with if(is_array()) ... , too. I presumed it would return only a value because there is only one value matching the query. I only worked with mysqli->query, num_rows and fetch_assoc before, not with the serendipity implementation of the database handling. Because all other SELECT calls with serendipity_db_query use is_array, I changed that now.

**Comment by onli** [comment]
Okay. So the next step is to test this, it looks like it should work fine. I have a test installation and will try to do that soon.

**Comment by th-h** [comment]
With the recent commits, the PR affects not only `serendipity_event_spamblock`, but also `serendipity_event_nl2br`. I do not think that was really intended ...

**Comment by stephanbrunker** [comment]
I am quite new to git. This pull request it did in autumn was still open and my working copy of the repo had these commits around the spamblock. I use this changes since then in my own blog because I have a bunch of nasty spammers blocking some older articles. Then a new request came up to bugfix nl2br in between. I created a new branch and cherry-picked that commit, and this PR was merged into the master by onli yesterday. It may be that this PR (from my master) has now also the nl2br commit into it, but that doesn't matter as these are already merged as https://github.com/s9y/Serendipity/commit/04cc9d06996906168c29d860b525d4fdb119b8da

**Comment by onli** [comment]
I'm actually not sure that it does not matter, the file changes tab shows that merging this PR would change files in nl2br. Which is actually really strange if those change were already in the other PR. Would you be okay with opening a new PR with just the relevant changes? If you still don't have the resources to test the upgrader I'd then test it and merge it asap.

**Comment by stephanbrunker** [comment]
That should do it. I reverted my nl2br commit, and merged and rebased with origin (where the commit was already applied). So the difference are only these four commits in spamblock and upgrader.

**Comment by onli** [comment]
> My git cheatsheet is still growing ...

Well done though :) I wasn't sure how to solve this either.

The code looks very good. I will merge this first and test then, this should work. Could you add one commit more that bumps the version number of the spamblock plugin to 1.87? That way all the relevant code changes are included in this PR.

**Comment by onli** [comment]
Thanks!

**Comment by onli** [comment]
Okay. I tested the upgrade task now and in both configurations - active via a day limit or set to 0 to be inactive - the upgrade task worked as expected. The code works fine :)

However, it's more inconsistent now than I realized. You are right, it now works a bit like the captcha toggle. But we also have an option "Trackback/Pingbackmoderation nach wievielen Tagen erzwingen" where 0 still means deactivated. We also have the IP-Block Interval, where 0 means not to block. And "Erforderliche Anzahl an Links für Moderation" as well as "Erforderliche Anzahl an Links für Abweisung" also work with 0 being the value that deactivates that filter. I assume we have more options like that in other plugin configurations.

Basically, the only inconsistent option I see was the "Captchas nach wie vielen Tagen erzwingen" option, which is the one this PR is modeled after :/ And that option is special because the toggle above also swaps between the captcha types.

---

So I'm now thinking: Maybe we should restore how it worked before and use the code to fix the captcha setting. Or alternatively, do we maybe need a configuration type "intandbool" that creates a widget where you can enter a number or also control whether the option is being active with a checkbox? Hierarchic menu options?
